sig-and-wg/sig-core/proposals/objects-store-proposal.md

```diff
@@ -0,0 +1,155 @@
+# Objects Store
+
+Created on 2019-01-03 by Lukasz Gornicki (@derberg).
+
+## Status
+
+Proposed on 2019-01-04
+
+## Motivation
+
+1. Have a generic solution for storing in s3 like storage any kind of object: zip file, markdown file, PNG or JS (JavaScript)
+2. Have a storage solution not locked to one vendor, like AWS S3 only, or like to Minio on-premise only
+3. Have a storage solution that not only stores the file but also exposes it directly to internet
+
+## Use case
+
+1. Storage for documentation and related images
+2. Storege for API specifications
+3. Storage for static client-side applications
+
+## Solution
+  
+![](assets/storage.svg)
+
+1. Location of the object is specified in the Object custom resource.
+2. ObjectStore controller fetches the object basing on the information given in the custom resource.
+3. The controller performs:
+    - Mutation of the object by communicating with mutation webhook specified in the custom resource
```

Comment:
My understanding is that a mutating webhook works by registering the mutating webhook per CRD and k8s calls the webhook on CRD events directly. I believe the controller does not communicate with the webhook.

Reply:
it is not about the MutatingWebhook I think you are referring to (https://github.com/morvencao/kube-mutating-webhook-tutorial/blob/master/deployment/mutatingwebhook.yaml) as this one enables you to do a mutation of the CRD. Here we need a custom approach here, where we can enable mutation of the fetched object before it is pushed to the bucket, to enable things like rewriting, unziping and things like that. You will see more explanation once I'm done